File: C++/Medium/RuntimeError/check-knight-tour-configuration_9.cpp
class Solution {
public:
    bool checkValidGrid(vector<vector<int>>& grid) {
        
        int N = grid.size();
        int M = grid[0].size();
        int move[8][2] = {{2,1},{2,-1},{-2,1},{-2,-1},{1,2},{1,-2},{-1,2},{-1,-2}};
        int row = 0, col = 0, count = 0;
        while (count < N*M) {
            int next_row = row + move[grid[row][col]][0];
            int next_col = col + move[grid[row][col]][1];
            if (next_row < 0 || next_row >= N || next_col < 0 || next_col >= M) {
                return false;
            }
            row = next_row;
            col = next_col;
            count++;
        }
        return true;
    }
    bool isValid(int row, int col, vector<vector<int>>& grid) {
        int N = grid.size();
        int M = grid[0].size();
        return row >= 0 && row < N && col >= 0 && col < M;
    }
    bool checkValidGrid2(vector<vector<int>>& grid) {
        int N = grid.size();
        int M = grid[0].size();
        int move[8][2] = {{2,1},{2,-1},{-2,1},{-2,-1},{1,2},{1,-2},{-1,2},{-1,-2}};
        vector<vector<bool>> visited(N, vector<bool>(M, false));
        int row = 0, col = 0, count = 0;
        while (count < N*M) {
            if (visited[row][col]) {
                return false;
            }
            visited[row][col] = true;
            int next_row = row + move[grid[row][col]][0];
            int next_col = col + move[grid[row][col]][1];
            if (!isValid(next_row, next_col, grid)) {
                return false;
            }
            row = next_row;
            col = next_col;
            count++;
        }
        return true;
    }
};

File: Java/Medium/WrongAnswer/check-knight-tour-configuration_9.java
class Solution {
    public boolean checkValidGrid(int[][] grid) {
        
        int n = grid.length;
        int[] dirs = {1, 2, 2, 1, -1, -2, -2, -1};
        boolean[][] visited = new boolean[n][n];
        int i = 0, j = 0;
        for (int k = 0; k < n * n; k++) {
            if (grid[i][j] != k) {
                return false;
            }
            visited[i][j] = true;
            for (int d = 0; d < 8; d += 2) {
                int x = i + dirs[d];
                int y = j + dirs[d + 1];
                if (x >= 0 && x < n && y >= 0 && y < n && !visited[x][y]) {
                    i = x;
                    j = y;
                    break;
                }
            }
        }
        return true;
    }
}